### PR TITLE
feat(imageviewer): optimizations to prefetch

### PR DESCRIPTION
**Pull request by ahorowitz123**
## Summary

This PR enhances the ImageViewer's loading performance and reliability by optimizing image completion detection and improving the prefetch functionality. The changes reduce loading times for already-cached images and make the prefetch system more robust with better error handling and comprehensive test coverage.

### 🚀 Performance Improvements

- **Immediate Loading Detection**: Added logic to `handleAssetAndRepLoad()` to immediately call `finishLoading()` if an image is already complete (cached), eliminating unnecessary wait times
- **Enhanced Prefetch System**: Extended the `prefetch()` method to support a new `preload` option alongside the existing `content` option for more flexible prefetching strategies
- **Browser Memory Caching**: Prefetched images are now rendered in the DOM (hidden off-screen) which forces the browser to decode and cache the image in memory, significantly improving subsequent load times when the same image is displayed

### 🛡️ Robustness & Error Handling  

- **Null-Safe Event Handling**: Added optional chaining (`?.`) to `prefetchFinishedLoading()` to prevent crashes when event targets are undefined
- **Default Parameter Fix**: Fixed parameter destructuring in `prefetch()` method to handle calls without arguments by adding default empty object (`= {}`)
- **Preview Mask Management**: Added `hidePreviewMask()` call to the overridden `finishLoading()` method for proper UI state management
- **Automatic Cleanup**: Prefetched images are automatically removed from the DOM after loading to prevent memory leaks

### 🧪 Comprehensive Testing

- **91 Total Tests**: Expanded test suite with 15+ new test cases covering all new functionality
- **Parameterized Testing**: Implemented `test.each` for efficient testing of multiple input combinations
- **DOM Testing**: Added comprehensive DOM manipulation and event handling tests
- **Edge Case Coverage**: Tests for watermarked files, incomplete representations, and error conditions
- **Modern Jest Practices**: Updated all assertions to use `toHaveBeenCalled()` instead of deprecated `toBeCalled()`

### Modified Files

- **`src/lib/viewers/image/ImageViewer.js`**
  - Enhanced `handleAssetAndRepLoad()` with immediate completion check
  - Improved `prefetch()` with `preload` option and default parameters
  - Added null-safe event handling in `prefetchFinishedLoading()`
  - Overridden `finishLoading()` to include preview mask management

- **`src/lib/viewers/image/__tests__/ImageViewer-test.js`**
  - Added comprehensive test coverage for all new functionality
  - Implemented parameterized tests for prefetch combinations
  - Added DOM cleanup hooks (`beforeEach`/`afterEach`)
  - Updated deprecated Jest assertions
  - Fixed prototype mocking to use `Object.getPrototypeOf()`

### Prefetch Memory Caching Strategy

The enhanced prefetch system now creates actual `` elements in the DOM to leverage browser optimizations:

```javascript
// Creates hidden image element that forces browser caching
const preFetchedImg = document.createElement('img');
preFetchedImg.addEventListener('load', this.prefetchFinishedLoading);
preFetchedImg.classList.add(CLASS_PREFETCHED_IMAGE); // Hidden off-screen via CSS
document.body.appendChild(preFetchedImg);
preFetchedImg.src = imageUrl; // Triggers download and browser memory caching
```

This approach ensures that:
- Images are fully decoded and stored in the browser's memory cache
- Subsequent loads of the same image are near-instantaneous
- The browser's native caching mechanisms are fully utilized
- Memory is properly cleaned up after prefetch completion

### New Test Coverage

- ✅ Image completion detection in `handleAssetAndRepLoad()`
- ✅ Preview mask hiding in `finishLoading()`
- ✅ DOM cleanup in `prefetchFinishedLoading()`
- ✅ Prefetch parameter combinations (`content`, `preload`)
- ✅ Prefetch edge cases (watermarked files, incomplete representations)
- ✅ DOM manipulation and event listener attachment
- ✅ Default parameter handling

## Performance Impact

- **Faster Loading**: Images that are already cached/complete now load immediately without waiting for load events
- **Better UX**: Reduced perceived loading times for frequently accessed images
- **Memory Optimization**: Prefetched images are stored in browser memory cache through DOM rendering, making subsequent loads nearly instantaneous
- **Robust Prefetching**: More reliable prefetch system with better error handling and automatic cleanup

## Testing

All 91 tests pass, including:
- Existing functionality regression tests
- New feature tests with edge case coverage
- DOM manipulation and cleanup verification
- Event handling and error condition tests

```bash
npm test -- src/lib/viewers/image/__tests__/ImageViewer-test.js
# ✅ Test Suites: 1 passed, 1 total
# ✅ Tests: 91 passed, 91 total
```

## Backward Compatibility

✅ All changes are backward compatible. Existing `prefetch()` calls continue to work as before, with new functionality available via optional parameters.

### CODE REVIEW

File: src/lib/viewers/BaseViewer.js [hunk 1]
```diff
@@ -1400,6 +1400,30 @@ class BaseViewer extends EventEmitter {
...
         return isFeatureEnabled(this.options.features, feature);
     }
+
+    /**
+     * Hides the preview mask element if it exists
+     *
+     * @return {void}
+     */
+    hidePreviewMask() {
+        const previewMask = document.getElementsByClassName('bcpr-PreviewMask')[0];
```
Comment (JChan106) [inline src/lib/viewers/BaseViewer.js]: Should we optional chain this? `document.getElementsByClassName('bcpr-PreviewMask')?[0]`

File: src/lib/viewers/BaseViewer.js [hunk 2]
```diff
@@ -1400,6 +1400,30 @@ class BaseViewer extends EventEmitter {
...
+        }
+    }
+
+    /**
+     * Shows the preview mask element if it exists
+     *
+     * @return {void}
+     */
+    showPreviewMask() {
+        const previewMask = document.getElementsByClassName('bcpr-PreviewMask')[0];
```
Comment (JChan106) [inline src/lib/viewers/BaseViewer.js]: ^

File: src/lib/viewers/image/Image.scss [hunk 3]
```diff
@@ -56,3 +56,11 @@
         }
     }
 }
+
+.bp-prefetched-image {
+    position: absolute;
+    left: -9999px;
```
Comment (JChan106) [inline src/lib/viewers/image/Image.scss]: What's the reason for the -9999px? Do we need comments around this?

Comment (ahorowitz123) [reply src/lib/viewers/image/Image.scss]: It's just make sure its completely offscreen so the user doesn't see it. It doesn't really need to be this high but its just exaggerated to ensure its not visible

Comment (JChan106) [reply src/lib/viewers/image/Image.scss]: would `display: none` work here instead?

Comment (ahorowitz123) [reply src/lib/viewers/image/Image.scss]: no because it has to actually render on the screen to have it stored in the browser memory cache

Comment (jmcbgaston) [reply src/lib/viewers/image/Image.scss]: couldn't we use `visibility: hidden`?

https://developer.mozilla.org/en-US/docs/Web/CSS/visibility

**Review by jmcbgaston — APPROVED** [review]

**Review by JChan106 — APPROVED** [review]
